File: src/test/java/com/contentful/java/cda/lib/TestCallback.java
package com.contentful.java.cda.lib;

import com.contentful.java.cda.CDACallback;
import com.contentful.java.cda.CDAResource;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public class TestCallback<T extends CDAResource> extends CDACallback<T> {
  private T result;

  private Throwable error;

  final CountDownLatch latch = new CountDownLatch(1);

  @Override protected void onSuccess(T result) {
    this.result = result;
    latch.countDown();
  }

  @Override protected void onFailure(Throwable error) {
    this.error = error;
    latch.countDown();
  }

  public TestCallback<T> await() throws InterruptedException {
    latch.await(1, TimeUnit.SECONDS);
    return this;
  }

  public T result() {
    return result;
  }

  public Throwable error() {
    return error;
  }
}

File: src/test/kotlin/com/contentful/java/cda/lib/TestCallback.kt
/*
 * Copyright (C) 2014 Contentful GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.contentful.java.cda.lib

import java.util.concurrent.CountDownLatch
import retrofit.RetrofitError
import com.contentful.java.cda.CDACallback

/**
 * TestCallback.
 */
class TestCallback<T>(val allowEmpty: Boolean = false) : CDACallback<T>() {
    val cdl: CountDownLatch
    var value: T = null
    var error: RetrofitError? = null

    init {
        cdl = CountDownLatch(1)
    }

    override fun onSuccess(result: T) {
        value = result
        cdl.countDown()
    }

    override fun onFailure(retrofitError: RetrofitError?) {
        super<CDACallback>.onFailure(retrofitError)
        error = retrofitError
        cdl.countDown()
    }

    throws(javaClass<InterruptedException>())
    public fun await() {
        cdl.await()
    }
}